File: tests/sdk/go/vpcflow_inject_test.go
package sdk_test

import (
	"context"
	"fmt"
	"io"
	"os"
	"strings"
	"testing"

	"github.com/aws/aws-sdk-go-v2/aws"
	"github.com/aws/aws-sdk-go-v2/service/s3"
)

func TestVPCFlowCreateInjectS3AcceptReject(t *testing.T) {
	requireReady(t)
	if os.Getenv("NOCTAXRIS_VPCFLOW_INJECT") != "1" {
		t.Skip("set NOCTAXRIS_VPCFLOW_INJECT=1 on the API process for lab InjectFlowLogs")
	}

	cfg := loadAWSConfig(t)
	s3c := newS3(t, cfg)
	ctx := context.Background()
	prefix := uniquePrefix(t)
	bucket := strings.ToLower(prefix + "-vpcflow")

	if _, err := s3c.CreateBucket(ctx, &s3.CreateBucketInput{Bucket: aws.String(bucket)}); err != nil {
		t.Fatalf("CreateBucket: %v", err)
	}
	t.Cleanup(func() {
		listed, _ := s3c.ListObjectsV2(ctx, &s3.ListObjectsV2Input{Bucket: aws.String(bucket)})
		for _, obj := range listed.Contents {
			_, _ = s3c.DeleteObject(ctx, &s3.DeleteObjectInput{Bucket: aws.String(bucket), Key: obj.Key})
		}
		_, _ = s3c.DeleteBucket(ctx, &s3.DeleteBucketInput{Bucket: aws.String(bucket)})
	})

	createStatus, createBody, createParsed := signedJSONTarget(t, "ec2", "AmazonEC2.CreateFlowLogs", map[string]any{
		"ResourceIds":        []string{"vpc-labopaque001"},
		"ResourceType":       "VPC",
		"TrafficType":        "ALL",
		"LogDestinationType": "s3",
		"LogDestination":     fmt.Sprintf("arn:aws:s3:::%s", bucket),
	})
	if createStatus != 200 {
		t.Fatalf("CreateFlowLogs status=%d body=%s", createStatus, createBody)
	}
	flowLogIDs, _ := createParsed["FlowLogIds"].([]any)
	if len(flowLogIDs) != 1 {
		t.Fatalf("CreateFlowLogs FlowLogIds=%v", createParsed)
	}
	flowLogID, _ := flowLogIDs[0].(string)
	if !strings.HasPrefix(flowLogID, "fl-") {
		t.Fatalf("FlowLogId=%q want fl- prefix", flowLogID)
	}

	injStatus, injBody, _ := signedJSONTarget(t, "ec2", "NoctaxrisEC2.InjectFlowLogs", map[string]any{
		"FlowLogId": flowLogID,
	})
	if injStatus != 200 {
		t.Fatalf("InjectFlowLogs status=%d body=%s", injStatus, injBody)
	}

	listed, err := s3c.ListObjectsV2(ctx, &s3.ListObjectsV2Input{
		Bucket: aws.String(bucket),
		Prefix: aws.String("AWSLogs/"),
	})
	if err != nil {
		t.Fatalf("ListObjectsV2: %v", err)
	}
	if len(listed.Contents) == 0 {
		t.Fatal("no flow log object in S3")
	}
	obj, err := s3c.GetObject(ctx, &s3.GetObjectInput{
		Bucket: aws.String(bucket),
		Key:    listed.Contents[0].Key,
	})
	if err != nil {
		t.Fatalf("GetObject: %v", err)
	}
	defer obj.Body.Close()
	raw, err := io.ReadAll(obj.Body)
	if err != nil {
		t.Fatalf("read object: %v", err)
	}
	text := string(raw)
	if !strings.Contains(text, " ACCEPT OK") || !strings.Contains(text, " REJECT OK") {
		t.Fatalf("s3 body=%q want ACCEPT+REJECT v2 lines", text)
	}
}

File: tests/sdk/nodejs/test/vpcflow-inject.test.mjs
import { test } from "node:test";
import assert from "node:assert/strict";
import {
  CreateBucketCommand,
  DeleteBucketCommand,
  DeleteObjectCommand,
  GetObjectCommand,
  ListObjectsV2Command,
} from "@aws-sdk/client-s3";
import {
  newS3,
  requireReady,
  signedJsonTarget,
  uniquePrefix,
} from "../lib/helpers.mjs";

async function streamToString(body) {
  if (!body) return "";
  if (typeof body.transformToString === "function") {
    return body.transformToString();
  }
  const chunks = [];
  for await (const chunk of body) {
    chunks.push(chunk);
  }
  return Buffer.concat(chunks).toString("utf8");
}

test("VPC Flow CreateFlowLogs InjectFlowLogs S3 ACCEPT REJECT", async (t) => {
  if (!(await requireReady(t))) return;
  if (process.env.NOCTAXRIS_VPCFLOW_INJECT !== "1") {
    t.skip(
      "set NOCTAXRIS_VPCFLOW_INJECT=1 on the API process for lab InjectFlowLogs",
    );
    return;
  }

  const s3 = newS3();
  const prefix = uniquePrefix();
  const bucket = `${prefix}-vpcflow`.toLowerCase();

  await s3.send(new CreateBucketCommand({ Bucket: bucket }));
  t.after(async () => {
    try {
      const listed = await s3.send(
        new ListObjectsV2Command({ Bucket: bucket }),
      );
      for (const obj of listed.Contents || []) {
        await s3.send(
          new DeleteObjectCommand({ Bucket: bucket, Key: obj.Key }),
        );
      }
    } catch {
      // best-effort cleanup
    }
    try {
      await s3.send(new DeleteBucketCommand({ Bucket: bucket }));
    } catch {
      // best-effort cleanup
    }
  });

  const created = await signedJsonTarget("ec2", "AmazonEC2.CreateFlowLogs", {
    ResourceIds: ["vpc-labopaque001"],
    ResourceType: "VPC",
    TrafficType: "ALL",
    LogDestinationType: "s3",
    LogDestination: `arn:aws:s3:::${bucket}`,
  });
  assert.equal(
    created.status,
    200,
    `CreateFlowLogs status=${created.status} body=${created.body}`,
  );
  const flowLogIds = created.json?.FlowLogIds || [];
  assert.equal(flowLogIds.length, 1);
  const flowLogId = flowLogIds[0];
  assert.ok(
    String(flowLogId).startsWith("fl-"),
    `FlowLogId=${flowLogId} want fl- prefix`,
  );

  const injected = await signedJsonTarget(
    "ec2",
    "NoctaxrisEC2.InjectFlowLogs",
    { FlowLogId: flowLogId },
  );
  assert.equal(
    injected.status,
    200,
    `InjectFlowLogs status=${injected.status} body=${injected.body}`,
  );

  const listed = await s3.send(
    new ListObjectsV2Command({ Bucket: bucket, Prefix: "AWSLogs/" }),
  );
  assert.ok((listed.Contents || []).length > 0, "no flow log object in S3");
  const key = listed.Contents[0].Key;
  const obj = await s3.send(
    new GetObjectCommand({ Bucket: bucket, Key: key }),
  );
  const text = await streamToString(obj.Body);
  assert.ok(text.includes(" ACCEPT OK"), `s3 body missing ACCEPT: ${text}`);
  assert.ok(text.includes(" REJECT OK"), `s3 body missing REJECT: ${text}`);
});
